src/main/java/org/prebid/server/auction/gpp/AmpGppService.java

```diff
@@ -0,0 +1,105 @@
+package org.prebid.server.auction.gpp;
+
+import com.iab.openrtb.request.BidRequest;
+import com.iab.openrtb.request.Regs;
+import com.iab.openrtb.request.User;
+import org.apache.commons.lang3.ObjectUtils;
+import org.prebid.server.auction.gpp.model.GppContext;
+import org.prebid.server.auction.gpp.model.GppContextCreator;
+import org.prebid.server.auction.model.AuctionContext;
+import org.prebid.server.util.ObjectUtil;
+
+import java.util.List;
+import java.util.Optional;
+
+public class AmpGppService {
+
+    private final GppService gppService;
+
+    public AmpGppService(GppService gppService) {
+        this.gppService = gppService;
+    }
+
+    public BidRequest apply(BidRequest bidRequest, AuctionContext auctionContext) {
+        final GppContext gppContext = gppService.processContext(contextFrom(bidRequest));
+
+        updateAuctionContext(auctionContext, gppContext);
+        return updateBidRequest(bidRequest, gppContext);
+    }
+
+    private static GppContext contextFrom(BidRequest bidRequest) {
+        final Regs regs = bidRequest.getRegs();
+
+        final String gpp = regs != null ? regs.getGpp() : null;
+        final List<Integer> gppSid = regs != null ? regs.getGppSid() : null;
+
+        final Integer gdpr = regs != null ? regs.getGdpr() : null;
+        final String consent = ObjectUtil.getIfNotNull(bidRequest.getUser(), User::getConsent);
+
+        final String usPrivacy = regs != null ? regs.getUsPrivacy() : null;
+
+        return GppContextCreator.from(gpp, gppSid)
+                .withTcfEuV2(gdpr, consent)
+                .withUspV1(usPrivacy)
+                .build();
+    }
+
+    private static void updateAuctionContext(AuctionContext auctionContext, GppContext gppContext) {
+        auctionContext.getDebugWarnings().addAll(gppContext.getErrors());
+    }
+
+    private static BidRequest updateBidRequest(BidRequest bidRequest, GppContext gppContext) {
+        final User user = bidRequest.getUser();
+        final User updatedUser = updateUser(user, gppContext);
+
+        final Regs regs = bidRequest.getRegs();
+        final Regs updatedRegs = updateRegs(regs, gppContext);
+
+        return ObjectUtils.anyNotNull(updatedUser, updatedRegs)
+                ? bidRequest.toBuilder()
+                .user(ObjectUtils.defaultIfNull(updatedUser, user))
+                .regs(ObjectUtils.defaultIfNull(updatedRegs, regs))
+                .build()
+                : bidRequest;
+    }
+
+    private static User updateUser(User user, GppContext gppContext) {
```

Comment:
Think about utilizing UpdateResult model, maybe it will be more logical then returning null.